File: connection-backend/src/main/java/de/florianschmitt/service/util/DTOMapper.java
package de.florianschmitt.service.util;

import de.florianschmitt.model.entities.*;
import de.florianschmitt.model.rest.*;
import de.florianschmitt.repository.LanguageRepository;
import lombok.NonNull;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.convert.converter.Converter;
import org.springframework.data.domain.Page;
import org.springframework.data.jpa.domain.AbstractPersistable;
import org.springframework.stereotype.Service;

import java.util.Collection;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class DTOMapper {

    @Autowired
    private ModelMapper mapper;

    @Autowired
    private LanguageRepository languageRepository;

    private EVoucherDTO map(EVoucher source) {
        EVoucherDTO result = mapper.map(source, EVoucherDTO.class);
        result.setVolunteerString(source.getVolunteer().toDisplayString());
        result.setRequestIdentifier(source.getRequest().getRequestIdentifier());
        return result;
    }

    private EPaymentDTO map(EPayment source) {
        EPaymentDTO result = mapper.map(source, EPaymentDTO.class);
        result.setPaymentBookedBy(source.getPaymentBookedBy().toDisplayString());
        result.setRequestId(source.getRequest().getRequestIdentifier());
        return result;
    }

    private ELanguageAdminDTO map(ELanguage source) {
        ELanguageAdminDTO result = mapper.map(source, ELanguageAdminDTO.class);

        List<ELocalizedDTO> list = source.getLocalized().stream()//
                .map(this::map)//
                .sorted(CustomELocalizedDTOComparator.INSTANCE)//
                .collect(Collectors.toList());

        result.setLocalized(list);

        return result;
    }

    private ELanguage map(ELanguageAdminDTO source) {
        ELanguage result = mapper.map(source, ELanguage.class);

        Set<ELocalized> col = source.getLocalized().stream()//
                .map(this::map)//
                .peek(loc -> loc.setLanguage(result))
                .collect(Collectors.toSet());
        result.setLocalized(col);

        return result;
    }

    private ELocalizedDTO map(ELocalized source) {
        ELocalizedDTO result = mapper.map(source, ELocalizedDTO.class);
        result.setLocale(source.getLocaleLanguage().name().toLowerCase());
        return result;
    }

    private ELocalized map(ELocalizedDTO source) {
        ELocalized result = mapper.map(source, ELocalized.class);
        result.setLocaleLanguage(ELocalizedLanguageEnum.Companion.create(source.getLocale()));
        return result;
    }

    private ERequest map(ERequestDTO source) {
        ERequest result = mapper.map(source, ERequest.class);
        result.setLanguages(source.getLanguageIds().stream()//
                .map(languageRepository::findOne)//
                .collect(Collectors.toSet()));
        return result;
    }

    private ERequestDTO map(ERequest source) {
        ERequestDTO result = mapper.map(source, ERequestDTO.class);
        result.setLanguageIds(source.getLanguages().stream()//
                .map(AbstractPersistable::getId)//
                .collect(Collectors.toSet()));
        if (source.getAcceptedByVolunteer() != null) {
            result.setAcceptedByVolunteer(source.getAcceptedByVolunteer().toDisplayString());
        }
        return result;
    }

    private EVolunteer map(EVolunteerDTO source) {
        EVolunteer result = mapper.map(source, EVolunteer.class);
        result.setLanguages(source.getLanguageIds().stream()//
                .map(languageRepository::findOne)//
                .collect(Collectors.toSet()));
        return result;
    }

    private EVolunteerDTO map(EVolunteer source) {
        EVolunteerDTO result = mapper.map(source, EVolunteerDTO.class);
        result.setLanguageIds(source.getLanguages().stream()//
                .map(AbstractPersistable::getId)//
                .collect(Collectors.toSet()));
        return result;
    }

    @SuppressWarnings("unchecked")
    public <D> D map(Object source, Class<D> destinationType) {
        if (EVolunteerDTO.class.equals(destinationType)) {
            return (D) map((EVolunteer) source);
        } else if (EVolunteer.class.equals(destinationType)) {
            return (D) map((EVolunteerDTO) source);
        } else if (ELanguageAdminDTO.class.equals(destinationType)) {
            return (D) map((ELanguage) source);
        } else if (EVoucherDTO.class.equals(destinationType)) {
            return (D) map((EVoucher) source);
        } else if (ELanguage.class.equals(destinationType) && ELanguageAdminDTO.class.isInstance(source)) {
            return (D) map((ELanguageAdminDTO) source);
        } else if (ERequestDTO.class.equals(destinationType)) {
            return (D) map((ERequest) source);
        } else if (ERequest.class.equals(destinationType)) {
            return (D) map((ERequestDTO) source);
        } else if (EPaymentDTO.class.equals(destinationType)) {
            return (D) map((EPayment) source);
        }
        D result = mapper.map(source, destinationType);
        return result;
    }

    public <D> Collection<D> map(@NonNull Collection<?> source, Class<D> destinationType) {
        return source.stream()//
                .map(x -> this.map(x, destinationType))//
                .collect(Collectors.toList());
    }

    public <D> List<D> map(@NonNull List<?> source, Class<D> destinationType) {
        return source.stream()//
                .map(x -> this.map(x, destinationType))//
                .collect(Collectors.toList());
    }

    public <D> Page<D> map(@NonNull Page<?> source, Class<D> destinationType) {
        return source.map(new Converter<Object, D>() {

            @Override
            public D convert(Object source) {
                return map(source, destinationType);
            }
        });
    }
}

File: src/main/kotlin/de/florianschmitt/service/util/DTOMapper.kt
package de.florianschmitt.service.util

import de.florianschmitt.model.entities.*
import de.florianschmitt.model.rest.*
import de.florianschmitt.repository.LanguageRepository
import org.modelmapper.ModelMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.stereotype.Service

@Service
class DTOMapper {

    @Autowired
    private lateinit var mapper: ModelMapper

    @Autowired
    private lateinit var languageRepository: LanguageRepository

    private fun map(source: EPayment): EPaymentDTO {
        val result = mapper.map(source, EPaymentDTO::class.java)
        result.paymentBookedBy = source.paymentBookedBy.displayString
        result.requestId = source.request.requestIdentifier
        return result
    }

    private fun map(source: ELanguage): ELanguageAdminDTO {
        val result = mapper.map(source, ELanguageAdminDTO::class.java)

        result.localized = source.localized
                ?.map { map(it) }
                ?.sortedWith(CustomELocalizedDTOComparator.INSTANCE)
                ?.toList()

        return result
    }

    private fun map(source: ELanguageAdminDTO): ELanguage {
        val result = mapper.map(source, ELanguage::class.java)

        result.localized = source.localized
                ?.map { map(it) }
                ?.onEach { it.language = result }
                ?.toSet()

        return result
    }

    private fun map(source: ELocalized): ELocalizedDTO {
        val result = mapper.map(source, ELocalizedDTO::class.java)
        result.locale = source.localeLanguage.name.toLowerCase()
        return result
    }

    private fun map(source: ELocalizedDTO): ELocalized {
        val result = mapper.map(source, ELocalized::class.java)
        result.localeLanguage = ELocalizedLanguageEnum.create(source.locale!!)
        return result
    }

    private fun map(source: ERequestDTO): ERequest {
        val result = mapper.map(source, ERequest::class.java)

        result.languages = source.languageIds
                ?.mapNotNull { languageRepository.findById(it).orElse(null) }
                ?.toMutableSet()

        return result
    }

    private fun map(source: ERequest): ERequestDTO {
        val result = mapper.map(source, ERequestDTO::class.java)
        result.languageIds = source.languages
                ?.map { it.id }
                ?.toSet()

        result.acceptedByVolunteer = source.acceptedByVolunteer?.displayString
        result.occasionString = source.occasion
        result.occasionEnum = null;

        return result
    }

    private fun map(source: EVolunteerDTO): EVolunteer {
        val result = mapper.map(source, EVolunteer::class.java)

        result.languages = source.languageIds
                ?.mapNotNull { languageRepository.findById(it).orElse(null) }
                ?.toMutableSet()

        return result
    }

    fun <D> map(source: Any, destinationType: Class<D>): D {
        return when (destinationType) {
            EVolunteerDTO::class.java -> (source as EVolunteer).toDto()
            EVolunteer::class.java -> map(source as EVolunteerDTO)
            ELanguageAdminDTO::class.java -> map(source as ELanguage)
            EVoucherDTO::class.java -> (source as EVoucher).toDto()
            ELanguage::class.java -> if (source is ELanguageAdminDTO) map(source) else mapper.map(source, destinationType)
            ERequestDTO::class.java -> map(source as ERequest)
            ERequest::class.java -> map(source as ERequestDTO)
            EPaymentDTO::class.java -> map(source as EPayment)
            else -> mapper.map(source, destinationType)
        } as D
    }

    fun <D> map(source: Collection<*>, destinationType: Class<D>) = source
            .filterNotNull()
            .map { this.map(it, destinationType) }
            .toList()

    fun <D> map(source: List<*>, destinationType: Class<D>) = map(source as Collection<*>, destinationType)

    fun <D> map(source: Page<*>, destinationType: Class<D>) = source.map { map(it as Any, destinationType) }
}

fun EVoucher.toDto() = EVoucherDTO(
        id = this.id,
        identifier = this.identifier,
        requestIdentifier = this.request.requestIdentifier,
        volunteerString = this.volunteer.displayString,
        answer = this.answer)

fun EVolunteer.toDto() = EVolunteerDTO(
        id = this.id,
        firstname = this.firstname,
        lastname = this.lastname,
        isActive = this.isActive,
        email = this.email,
        languageIds = this.languages?.map { it.id }?.toSet())

fun ESystemUser.toDto() = ESystemUserDTO(
        id = this.id,
        firstname = this.firstname,
        lastname = this.lastname,
        isActive = this.isActive,
        email = this.email,
        hasAdminRight = this.hasAdminRight,
        hasFinanceRight = this.hasFinanceRight)

fun ERequest.toSimpleDto(): ERequestSimpleDTO {
    var result = ERequestSimpleDTO()
    result.datetime = this.datetime;
    result.languageIds = this.languages?.map { it.id }?.toSet()
    result.city = this.city
    result.ocation = this.occasion
    result.postalCode = this.postalCode
    result.street = this.street
    return result
}